sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -52,76 +67,191 @@ def _get_component_hash(cls, component: Component) -> str:
+        Distinct dependent components not in current cache will be saved to a list.
+
+        :return: a tuple of (dict of nodes to resolve, list of cache contents to resolve)
         """
-        if self._hash_of_components_to_resolve:
-            self._resolve_components_without_in_memory_cache()
+        _components = list(map(lambda x: x._component, self._nodes_to_resolve))  # pylint: disable=protected-access
+        # we can do concurrent component in-memory hash calculation here
+        in_memory_component_hashes = map(self._get_in_memory_hash_for_component, _components)
 
-        for component_hash, nodes in self._nodes_to_apply.items():
-            for node in nodes:
-                resolution_result = self._component_resolution_cache[component_hash]
+        dict_of_nodes_to_resolve = defaultdict(list)
+        cache_contents_to_resolve: List[_CacheContent] = []
+        for node, component_hash in zip(self._nodes_to_resolve, in_memory_component_hashes):
+            dict_of_nodes_to_resolve[component_hash].append(node)
+            if component_hash not in self._cache:
+                cache_content = _CacheContent(
+                    component_ref=node._component,  # pylint: disable=protected-access
+                    in_memory_hash=component_hash,
+                )
+                self._cache[component_hash] = cache_content
+                cache_contents_to_resolve.append(cache_content)
+        self._nodes_to_resolve.clear()
+        return dict_of_nodes_to_resolve, cache_contents_to_resolve
+
+    @classmethod
+    def _resolve_cache_contents_from_disk(
+        cls,
+        cache_contents_to_resolve: List[_CacheContent]
+    ) -> List[_CacheContent]:
+        """Check on-disk cache to resolve cache contents in cache_contents_to_resolve and return
+        unresolved cache contents.
+        """
+        if not is_on_disk_cache_enabled():
+            return cache_contents_to_resolve[:]
+
+        # Note that we should recalculate the hash based on code for local cache, as
+        # we can't assume that the code folder won't change among dependency resolution
+        for cache_content in cache_contents_to_resolve:
+            cache_content.on_disk_hash = cls._get_on_disk_hash_for_component(
+                cache_content.component_ref,
+                cache_content.in_memory_hash
+            )
+
+        left_cache_contents_to_resolve = []
+        # need to dedup disk hash first if concurrent resolution is enabled
+        for cache_content in cache_contents_to_resolve:
+            cache_content.arm_id = cls._load_from_on_disk_cache(cache_content.on_disk_hash)
+            if not cache_content.arm_id:
+                left_cache_contents_to_resolve.append(cache_content)
+
+        return left_cache_contents_to_resolve
 
+    @classmethod
+    def _save_cache_contents_to_disk(cls, cache_contents_to_resolve: List[_CacheContent]):
```

Comment:
why not doing save while resolved one component?

Reply:
Hmm, I write it in this way to decouple logic of on-disk cache and in-memory cache, but seems no harm to save while resolved one component